app/controllers/image_proxy_controller.rb

```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+##
+# Controller for proxying HTTP images
+class ImageProxyController < ApplicationController
+  before_action :validate_token
+
+  def access
+    send_data proxied_response.body, type: proxied_response.content_type, disposition: 'inline'
+  end
+
+  private
+
+  ##
+  # To prevent outside applications from using this proxy
+  def validate_token
+    render plain: 'invalid token', status: :forbidden unless valid_authenticity_token?(session, image_token)
```

Comment:
Is this not as smart as I think, as are we going to invalidate browser cache on every image?